### ISSUE TITLE
慢哈希算法之密码安全存储——PBKDF2、bcrypt、scrypt、

### ISSUE THREAD

**Issue by bingoohuang** [bingoohuang/blog#187]

**Comment by bingoohuang** [comment]
# Scrypt算法

Scrypt算法由知名的FreeBSD黑客，同时也是一名密码学家的Colin Percival在2009年提出，他开发这一算法的初衷是为其备份服务Tarsnap降低CPU负荷，减少对于CPU计算的依赖，同时防止网络攻击。虽然Scrypt算法的性能与算力相关不大，但却很依赖内存。

由于Scrypt算法对内存非常依赖的设计非常适合用于对抗专业矿机，所以后来有很多区块链挖矿项目都使用这一算法，如狗狗币，来使得数字货币的分发更加分散。

Scrypt算法属于一种密钥导出（KDF）算法，这种算法非常适用于生成密钥，可以避免黑客低成本地生成大量密钥去试探密码，因此可以为备份服务提供高度安全的网络环境。

Scrypt算法使得加密币网络更加安全和去中心化，所以一经莱特币的使用，它就得到了广大山寨币的青睐，有很多人甚至认为这是最适合数字货币的算法。

但是，随着Scrypt算法的广泛使用，专用于这一算法的矿机也被生产出来，后来诞生的FPGA矿机也可以在运用这算法时进行大规模挖矿。

**Comment by bingoohuang** [comment]
[age](https://github.com/FiloSottile/age) 的现代化体现在密码学算法的选择上，age 使用 x25519 作为非对称加密算法，x25519 是 Curve25519 被设计用于密钥交换的曲线，是目前公认的最快的椭圆密码曲线，它还有个用做签名的 ed25519 的兄弟，二者的公私钥可以互相转换。

摘自[age：一个现代化的加密工具](https://github.com/islishude/blog/issues/220)

**Comment by bingoohuang** [comment]
# Go 中提供了 golang.org/x/crypto，其中包括常用的 KDF：

1. Argon2, golang.org/x/crypto/argon2。
2. Bcrypt, golang.org/x/crypto/bcrypt。
3. Scrypt, golang.org/x/crypto/scrypt。
4. PBKDF2, golang.org/x/crypto/pbkdf2。
5. HKDF, golang.org/x/crypto/hkdf。（HKDF = HMAC + KDF）

**Comment by bingoohuang** [comment]
# 什么是AEAD加密

Authenticated Encryption with Associated Data (AEAD) 是一种同时具备保密性，完整性和可认证性的加密形式。

AEAD 产生的原因很简单，单纯的对称加密算法，其解密步骤是无法确认密钥是否正确的。也就是说，加密后的数据可以用任何密钥执行解密运算，得到一组疑似原始数据，而不知道密钥是否是正确的，也不知道解密出来的原始数据是否正确。

常见的 AEAD 算法如下：

1. AES-128-GCM
2. AES-192-GCM
3. AES-256-GCM
4. ChaCha20-IETF-Poly1305
5. XChaCha20-IETF-Poly1305

在具备 AES 加速的 CPU（桌面，服务器）上，建议使用 AES-XXX-GCM 系列，移动设备建议使用 ChaCha20-IETF-Poly1305 系列。


1. [什么是AEAD加密](https://zhuanlan.zhihu.com/p/28566058)
2.

**Comment by bingoohuang** [comment]
# 常用的对称加密算法


算法名 | 优点 | 缺点
-- | -- | --
AES-128-CBC | 实现简单，运行速度快 | 无 MAC 功能
AES-128-GCM | 有 MAC 功能 | 实现复杂，速度较 CBC 慢
ChaCha20-Poly1305 | 运行速度快，适用于移动端 | 推出时间较短
RC4 | 实现简单，运行速度快 | 已经不安全

AES-GCM 是目前常用的分组加密算法，但是其有一个缺点就是计算量大，导致性能和电量开销比较大。为了解决这个问题，intel 推出了名为 AES NI（Advanced Encryption Standard new instructions）的 x86 指令拓展集，从硬件上提供对 AES 的支持，具体参见文档。

所以对于支持 AES NI 指令的设备来说，使用 AES-GCM 无疑是最佳选择。针对移动端，对于不支持 AES NI 的设备来说，Google 在 2014年 推出了一种新的流式加密算法 ChaCha20-Poly1305，在 ARM 平台上，性能是 AES-GCM 的 3-4 倍。

总之，ChaCha20-Poly1305 作为一个加密组合，可同时对数据提供机密性，完整性和真实性保证，避开了现有发现的所有安全漏洞和攻击，是一组极佳的加密套件组合。

## 资料

1. [对称加密算法优化之 ChaCha20-Poly1305 算法支持](https://blog.upyun.com/?p=1704)

**Comment by bingoohuang** [comment]
[bcrypt vs pbkdf2 for encrypting private keys](https://stackoverflow.com/a/40642896)

You should not be using the bcrypt hash output as an encryption key; it is not meant to be key material:

- BCrypt is not a key-derivation function
- BCrypt it is a password storage function

You have an elliptic curve private key that you want to encrypt using a user's password. Of course you don't want to use the password directly - you want to use the password to derive an encryption key. For that you can use:

- PBKDF2
- scrypt

These are both key-derivation functions (e.g. password-based key derivation function). Their purpose is to generate an encryption key given a password. They are designed to be "hard".

You feed both these algorithms:

- a password
- cost parameters
- salt
- desired number of bytes (e.g. 32 ==> 32 bytes ==> 256 bits)

and it returns you a 256-bit key you can use as an encryption key to AES-256.

## You then want to backup the user's key

I gather that you then want to:

- store the encrypted elliptic curve private key on your server
- store a hash of their password on your server

And your question was: since you already ran their password through "a hashing function" can't you just use that hash as their stored password?

No! That hash is also the encryption key protecting their private key. You don't want that private key transmitted anywhere. You don't want it existing anywhere. That 32-byte encryption key should be wiped from memory as soon as you're done with it.

What you should do, if you also wish to store a hash of the user's password is use an algorithm that is typically used for password storage:

- pbkdf2 (a key-derivation function abused into password storage)
- bcrypt (better than pbkdf2)
- scrypt (a key-derivation function abused into password storage; better than bcrypt)
- argon2 (better than scrypt)

You should separately run the user's password through one of these password storage algorithms. If you have access to bcrypt; use that over pbkdf2. If you have scrypt, use that for both:

derivation of an encryption key
hashing of the password
The security of your system comes from (in addition to the secrecy of the password), the computational distance between the user's password and the encryption key protecting their private key:

- "hunter2"  --PBKDF2--> Key material
- "hunter2"  ---------bcrypt-------> Key material
- "hunter2"  ----------------scrypt----------> Key material

You want as much distance between the password and the key.

## Not-recommended cheat

If you're really desperate to save CPU cycles (and avoid computing scrypt twice), you technically could take:

Key Material ---SHA2---> "hashed password"

And call the hash of the encryption key your "hashed password" and store that. Computation of a single SHA2 is negligible. This is acceptable because the only way an attacker can use this is by trying to guess every possible 256-bit encryption key - which is the problem they can't solve in the first place. There's no way to bruteforce a 256-bit key. And if they were to try to brute-force it, the extra hashed version doesn't help them, as they could just test their attempt by trying to decrypt the private key.

But it's much less desirable because you're storing (a transformed) version of the encryption key. You want that key (and any transformed versions of it) stored as little as possible.

## To sum up

1. generate EC key pair
2. encryptionKey = scryptDeriveBytes(password, salt, cost, 32)
3. encryptedPrivateKey = AES256(privateKey, encryptionKey)
4. passwordHash = scryptHashPassword(password, salt, cost)

and upload

1. encryptedPrivateKey
2. passwordhash

**Comment by bingoohuang** [comment]
项目代码的构建扫描结果中，Sonarsource Rule 中建议用 AES - GCM 来替代 AES-CBC, 原因如下：

1. AES-GCM可以并行加密解密，AES-CBC的模式决定了它只能串行地进行加密。因为加密是耗时较久的步骤，且加密的方式是相同的，所以并行地实现AES-GCM算法的时候，其效率是高于AES-CBC的；
2. AES-GCM提供了GMAC信息校验码，用以校验密文的完整性。AES-CBC没有，无法有效地校验密文的完整性；
3. AES-GCM是流加密的模式，不需要对明文进行填充。AES-CBC是块加密的模式，需要对明文进行填充。(AES-GCM中进行AES加密的是counter，AES-CBC中进行AES加密的是明文块)；
4. 由于AES-CBC中必须要用到padding，导致最后一个明文块与其他密文块不同，因此可能会受到padding Oracle attacks，从而可以直接通过初始向量IV和密码，即可得到明文。

AES加密是对称加密的一种，即加密和解密使用相同的一把密钥。它的全称是Advanced Encryption Standard(高级加密标准)，主要是用来取代DES加密算法，目前已经被全世界广泛采用，各大处理器厂商也在各自的CPU中，集成了专门的AES指令集，从而在硬件层面提升了AES加解密的速度。

**Comment by bingoohuang** [comment]
# GCM：伽罗瓦计数器模式

类型CCM模式，GCM模式是CTR和GHASH的组合，GHASH操作定义为密文结果与密钥以及消息长度在GF（2^128）域上相乘。GCM比CCM的优势是在于更高并行度及更好的性能。*TLS 1.2标准使用的就是AES-GCM算法*，并且Intel CPU提供了GHASH的硬件加速功能。

来自[AES简介](https://github.com/matt-wu/AES)

![image](https://user-images.githubusercontent.com/1940588/122318417-968c2680-cf51-11eb-8240-f5bf634f4248.png)

GCM (Galois/Counter Mode) 是一种经过验证的加密演算法，除了保密性之外，还可确保数据完整性，因优异的效能而被广泛使用。Zoom 已在会议中使用ECB 从AES-256 加密转移到AES-256-GCM。

**Comment by bingoohuang** [comment]
> 加密分组对称加密算法还有一个分组模式 的概念，对于 GCM 分组模式，只有和 AES，CAMELLIA 和 ARIA 搭配使用，而 AES 显然是最受欢迎和部署最广泛的选择，它可以让算法用固定长度的密钥加密任意长度的明文。
*最早有 ECB、CBC、CFB、OFB 等几种分组模式，但都陆续被发现有安全漏洞*，所以现在基本都不怎么用了。最新的分组模式被称为 AEAD（Authenticated Encryption with Associated Data），在加密的同时增加了认证的功能，常用的是 GCM、CCM 和 Poly1305。
比如 ECDHE_ECDSA_AES128_GCM_SHA256 ，表示的是具有 128 位密钥， AES256 将表示 256 位密钥。GCM 表示具有 128 位块的分组密码的现代认证的关联数据加密（AEAD）操作模式。

[为什么浏览器会提示网站“不安全”？一文读懂https协议与SSL证书](https://bbs.huaweicloud.com/blogs/154560)

**Comment by bingoohuang** [comment]
# Curve25519 中的 25519 的含义

http://everything.explained.today/Curve25519/

Mathematical properties
The curve used is

y^2=x^3+486662x^2+x, a Montgomery curve, over the prime field defined by the prime number **2^255-19**, and it uses the base point

**Comment by bingoohuang** [comment]
# 密钥交换算法

http://z-rui.github.io/post/2020/12/x25519/


使用对称密钥通信时，双方需要事先约定密钥。如何安全地操作是一个关键问题。Diffie-Hellman-Merkle算法使得在不安全的信道中交换密钥成为可能。

在Diffie-Hellman-Merkle算法中，事先约定整数G和M。通信双方Alice和Bob分别拥有自己的私钥a和b。Alice把自己的公钥A=Ga发送给Bob。同理，Bob把B=Gb发送给Alice。Alice计算Ba，Bob计算Ab。这样，他们就得到了共享密钥K=Gab。

注意，这里的乘法定义在GF(M)上，即A=Ga mod M，等等。对于取模的指数运算，其逆运算是一个目前难以解决的问题。对于外界的观察者而言，只知道G、M、A、B，并不能容易地求出a、b或K。这样就确保了密钥交换过程的安全性。

在实践中，常用密钥派生函数KDF对密钥K进行处理，用派生密钥KDF(K)作为对称加密的实际密钥。

## X25519 密钥交换算法

X25519是一种类似Diffie-Hellman-Merkle，但基于椭圆曲线理论的密钥交换算法。

![image](https://user-images.githubusercontent.com/1940588/125816961-a1c9a989-a722-4550-ade1-1761bfeaae51.png)

**Comment by bingoohuang** [comment]
https://www.wolframalpha.com/input/?i=y%5E2%3Dx%5E3%2B486662x%5E2%2Bx

![image](https://user-images.githubusercontent.com/1940588/125888153-8c9c9fb8-a8ce-4646-9924-1c2ec5565d92.png)


https://www.wolframalpha.com/input/?i=Is+2%5E255-19+prime%3F

![image](https://user-images.githubusercontent.com/1940588/125888280-c3f1597e-d782-4bc9-b55c-c0dee30dc1cf.png)

**Comment by bingoohuang** [comment]
# [cocert](https://github.com/Dentrax/cocert)

cocert, generates [ECDSA](https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm) - [P521](https://en.wikipedia.org/wiki/Elliptic-curve_cryptography) key and uses a technique known as [Shamir's Secret Sharing](https://en.wikipedia.org/wiki/Shamir's_Secret_Sharing) algorithm to split the master key into x shares, any y of which are required to reconstruct the master private key. Private keys are stored in [PEM](https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail)-encoded [PKCS8](https://en.wikipedia.org/wiki/PKCS_8) format, which are [encrypted](https://pkg.go.dev/github.com/theupdateframework/go-tuf/encrypted) by [The Update Framework (TUF)](https://github.com/theupdateframework/go-tuf). Each private key is splitted using [Shamir Split](https://pkg.go.dev/github.com/hashicorp/vault/shamir#Split). To [Combine](https://pkg.go.dev/github.com/hashicorp/vault/shamir#Combine) private key files into single one, it is necessary to enter decrypt password if it has been encrypted by TUF.

**Comment by bingoohuang** [comment]
# 快 Hash

[Use Fast Data Algorithms](https://jolynch.github.io/posts/use_fast_data_algorithms/)

*TLDR*

Application | Common Bad Performance Choices | Better Performance Choices | Expected Performance Gain
-- | -- | -- | --
Trusted data hashing | md5, sha2, crc32 | [xxhash](http://cyan4973.github.io/xxHash/) | ~10x
Untrusted data hashing | md5, sha2, sha1 | [blake3](https://github.com/BLAKE3-team/BLAKE3) | ~10x
Fast compression | snappy, gzip (zlib) | [lz4](https://github.com/lz4/lz4) | 10x over gzip, ~2x over snappy
Good compression | gzip (zlib) | [zstd](https://facebook.github.io/zstd/) | ~2-10x
Best compression | xz (lzma) | zstd -10+ | ~2-10x
Java crypto (md5, aes-gcm, etc …) | [Built-in JVM crypto | Amazon Corretto Crypto Provider (ACCP)](https://github.com/corretto/amazon-corretto-crypto-provider) | ~4-10x

TODO 循着文章做一下自己的实锤验证

**Comment by bingoohuang** [comment]
from  README of https://github.com/dgrijalva/jwt-go

## Usage Tips

### Signing vs Encryption

A token is simply a JSON object that is signed by its author. this tells you exactly two things about the data:

* The author of the token was in the possession of the signing secret
* The data has not been modified since it was signed

It's important to know that JWT does not provide encryption, which means anyone who has access to the token can read its contents. If you need to protect (encrypt) the data, there is a companion spec, `JWE`, that provides this functionality. JWE is currently outside the scope of this library.

### Choosing a Signing Method

There are several signing methods available, and you should probably take the time to learn about the various options before choosing one.  The principal design decision is most likely going to be symmetric vs asymmetric.

Symmetric signing methods, such as HSA, use only a single secret. This is probably the simplest signing method to use since any `[]byte` can be used as a valid secret. They are also slightly computationally faster to use, though this rarely is enough to matter. Symmetric signing methods work the best when both producers and consumers of tokens are trusted, or even the same system. Since the same secret is used to both sign and validate tokens, you can't easily distribute the key for validation.

Asymmetric signing methods, such as RSA, use different keys for signing and verifying tokens. This makes it possible to produce tokens with a private key, and allow any consumer to access the public key for verification.

### Signing Methods and Key Types

Each signing method expects a different object type for its signing keys. See the package documentation for details. Here are the most common ones:

* The [HMAC signing method](https://godoc.org/github.com/dgrijalva/jwt-go#SigningMethodHMAC) (`HS256`,`HS384`,`HS512`) expect `[]byte` values for signing and validation
* The [RSA signing method](https://godoc.org/github.com/dgrijalva/jwt-go#SigningMethodRSA) (`RS256`,`RS384`,`RS512`) expect `*rsa.PrivateKey` for signing and `*rsa.PublicKey` for validation
* The [ECDSA signing method](https://godoc.org/github.com/dgrijalva/jwt-go#SigningMethodECDSA) (`ES256`,`ES384`,`ES512`) expect `*ecdsa.PrivateKey` for signing and `*ecdsa.PublicKey` for validation

### JWT and OAuth

It's worth mentioning that OAuth and JWT are not the same thing. A JWT token is simply a signed JSON object. It can be used anywhere such a thing is useful. There is some confusion, though, as JWT is the most common type of bearer token used in OAuth2 authentication.

Without going too far down the rabbit hole, here's a description of the interaction of these technologies:

* OAuth is a protocol for allowing an identity provider to be separate from the service a user is logging in to. For example, whenever you use Facebook to log into a different service (Yelp, Spotify, etc), you are using OAuth.
* OAuth defines several options for passing around authentication data. One popular method is called a "bearer token". A bearer token is simply a string that _should_ only be held by an authenticated user. Thus, simply presenting this token proves your identity. You can probably derive from here why a JWT might make a good bearer token.
* Because bearer tokens are used for authentication, it's important they're kept secret. This is why transactions that use bearer tokens typically happen over SSL.
